include/pybind11_json/pybind11_json.hpp

```diff
@@ -67,8 +68,13 @@ namespace pyjson
         }
     }
 
-    inline nl::json to_json(const py::handle& obj)
+    inline nl::json to_json(const py::handle& obj, std::set<const PyObject*>& refs)
     {
+        auto insert_ret = refs.insert(obj.ptr());
```

Comment:
There is one thing that makes me nervous about this approach, `id`s for numbers are the same, same for `True`, `False` and even strings in some cases (probably interpreter optimization). Would this logic wrongfully raise an error for these example lists:

![Screenshot From 2024-12-09 08-55-39](https://github.com/user-attachments/assets/bd9d54bc-afb1-4891-90c3-6fed9a57ab3f)

I guess we should only make this check for id for lists and dictionaries?

Reply:
I don't want to be asking too much, but adding a test for these would be great!

Reply:
I can also think of another case where we should be careful:

![Screenshot From 2024-12-09 09-00-21](https://github.com/user-attachments/assets/7dab0573-2dfb-402f-badd-690fb6a643d1)

Here there is no reason we can't serialize `a`, but I understand your logic here will raise an exception. We may need to make a set every time we pass it to the underlying dict/array? Only passing `id`s of parent objects? We want to make sure an object does not contain itself. But an object can contain twice the same object.

Reply:
Thanks for making changes! I'm still a bit nervous about this. Doesn't it raise in the case of serializing `b` in the following case, even though there is no circular reference?

```python
a = [1, 2, 3]
b = {
  "a1": a,
  "a2": a
}
```

What we want to check is that if an object contains itself, but we're checking if an object appears more than once.